### DISCUSSION TITLE
[Deployment] Deploying using Metamask Initialize provider L2

### DISCUSSION THREAD

**Discussion by 5208980** [zkSync-Community-Hub/zksync-developers#853] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

1.5.8

### zksync-ethers Version

6.15.3

### Minimal Reproducible Code or Repo Link

```ts
import { BrowserProvider, ContractFactory } from "zksync-ethers"

const provider = new BrowserProvider(window.ethereum)
const signer = await provider.getSigner()

const accounts: string[] = await window.ethereum.request({
  method: "eth_requestAccounts",
})

// Here the runner L2Provider is null
console.log(`ETH balance: ${await signer.getBalance()}`, signer);

const DEPLOYER_SYSTEM_CONTRACT_ADDRESS = "0x0000000000000000000000000000000000008006";

const factory = new ContractFactory(
  abi, bytecode,
  signer
);

const contract = await factory.deploy( ...args);

await contract.waitForDeployment();
```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Not sure

### Additional Details

I'm trying to deploy a smart contract on the ZkSync Sepolia Testnet. I've compiled it with zksolc and obtained the ABI and eravm bytecode from the output. However, when I try to deploy using zksync-ethers and Metamask (which is connected to the ZkSync Sepolia network), I get 'initialize provider L2'. This suggests that I might not have a valid L2Provider for my signers.

I've reviewed code snippets online, and most examples utilize `Wallet`, which allows passing a Provider from both `ethers` and `zksync-ethers`. Is there a way to deploy using the Metamask signer?

**Comment by alhenaai** [comment]
It seems like you're encountering an issue with initializing the L2 provider when trying to deploy a smart contract on the ZKsync Sepolia Testnet using MetaMask and `zksync-ethers`. The error 'initialize provider L2' typically indicates that the L2 context for the provider is not properly set up.

In your case, using `BrowserProvider` from `zksync-ethers` directly with MetaMask might not be fully compatible or may require additional configuration to properly establish the L2 context. Here are a few steps you can take to resolve this issue:

1. **Ensure Network Compatibility**: Verify that your MetaMask is correctly connected to the ZKsync Sepolia Testnet. This is crucial as the network settings directly affect the provider's ability to interact with L2.

2. **Use a Dedicated L2 Wallet**: Instead of using MetaMask directly, consider using a dedicated L2 wallet setup that can handle ZKsync-specific features. You can create a wallet instance using `zksync-ethers` that integrates both L1 and L2 providers. Here’s an example of how you might set it up:

   ```typescript
   import { ethers } from 'ethers';
   import { Wallet, Provider } from 'zksync-ethers';

   const l1Provider = new ethers.providers.Web3Provider(window.ethereum);
   const l2Provider = await Provider.getDefaultProvider('sepolia');

   const ethWallet = l1Provider.getSigner();
   const zkSyncWallet = await Wallet.fromEthSigner(ethWallet, l2Provider);
   ```

3. **Deploy Using the zkSyncWallet**: Once you have the `zkSyncWallet`, you can proceed with deploying your contract using the `ContractFactory` from `zksync-ethers`, similar to how you would with a regular ethers.js setup.

4. **Check Documentation and Examples**: Ensure that you are following the latest examples and documentation provided by ZKsync, which might have specific instructions or updated methods for deploying contracts using different setups. You can find more detailed guides and examples on the [ZKsync developer discussions on GitHub](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

If you continue to face issues, it might be helpful to look into the specific configurations or updates in the `zksync-ethers` library that pertain to handling L2 transactions with MetaMask.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sachanayush47** [reply]
Bot answer looks good to me.

**Reply by alhenaai** [reply]
Great to hear that the information was helpful! If you have any more questions or need further assistance as you continue working with ZKsync, feel free to reach out. Good luck with your deployment on the ZKsync Sepolia Testnet!

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by uF4No** [comment, marked as answer]
The error `initialize provider L2` is caused because you're using a default signer but you need to initialise the signer using the `Signer` class from `zksync-ethers`, which has special properties like two providers (one for L1, and one for L2).

I've updated your script:

```js
import { BrowserProvider, ContractFactory, Signer } from "zksync-ethers"

const provider = new BrowserProvider(window.ethereum)
const zkProvider = new Provider("https://sepolia.era.zksync.dev");

const signer = await provider.getSigner()

// I don't think you need this
const accounts: string[] = await window.ethereum.request({
  method: "eth_requestAccounts",
})

const chainId = Number((await provider.getNetwork()).chainId)
console.log("chain id >> ", chainId)

const zkSigner = Signer.from(signer, chainId, zkProvider);


// Here the runner L2Provider is null
console.log(`ETH balance: ${await zkSigner.getBalance()}`);

const DEPLOYER_SYSTEM_CONTRACT_ADDRESS = "0x0000000000000000000000000000000000008006";

const factory = new ContractFactory(
  abi, bytecode,
  signer
);

const contract = await factory.deploy( ...args);

await contract.waitForDeployment();
```

If you have other issues, make sure to check the `ContractFactory` API reference in the docs: https://docs.zksync.io/zksync-era/sdk/js/ethers/api/v6/contract/contract-factory
